### PR TITLE
feat: Predicate UDF Reordering Optimization

### PR DESCRIPTION
**Pull request by gaurav274**
1. Reorders the conjuncts based on the cost of the UDFs 
2. Bug fix: orderby executor did not safely handle corner case if the batch is empty.
3. Bug fix: TupleValue Expression did not respect the "mask" used by short-circuiting leading to wrong results.

### CODE REVIEW

File: eva/constants.py [hunk 1]
```diff
@@ -17,3 +17,4 @@
 NO_GPU = -1
 UNDEFINED_GROUP_ID = -1
 IFRAMES = "IFRAMES"
+DEFAULT_FUNCTION_EXPRESSION_COST = 100
```
Comment (jarulraj) [inline eva/constants.py:20]: In a future PR, we should look at why we have two `constants.py` files.

File: eva/optimizer/rules/rules.py [hunk 2]
```diff
@@ -499,6 +503,64 @@ def apply(self, before: LogicalJoin, context: OptimizerContext):
...
+
+    def apply(self, before: LogicalFilter, context: OptimizerContext):
+        # Decompose the expression tree into a list of conjuncts
+        conjuncts = to_conjunction_list(before.predicate)
+
+        # Segregate the conjuncts into simple and function expressions
+        contains_func_exprs = []
+        simple_exprs = []
+        for conjunct in conjuncts:
+            if list(conjunct.find_all(FunctionExpression)):
```
Comment (jarulraj) [inline eva/optimizer/rules/rules.py:535]: Do we have a test case where there are no `FunctionExpression` in `conjunct` -- edge case?

File: test/benchmark_tests/test_benchmark_pytorch.py [hunk 3]
```diff
@@ -80,7 +80,7 @@ def test_should_run_pytorch_and_resnet50(benchmark, setup_pytorch_tests):
     # non-trivial test case for Resnet50
     res = actual_batch.frames
     assert res["featureextractor.features"][0].shape == (1, 2048)
-    assert res["featureextractor.features"][0][0][0] > 0.3
+    # assert res["featureextractor.features"][0][0][0] > 0.3
```
Comment (jarulraj) [inline test/benchmark_tests/test_benchmark_pytorch.py:83]: why was this commented out?

File: test/integration_tests/test_optimizer_rules.py [hunk 4]
```diff
@@ -123,3 +125,96 @@ def test_should_pushdown_without_pushdown_join_rule(self):
...
         self.assertEqual(result_without_pushdown_join_rule, result_with_rule)
         self.assertEqual(query_plan, query_plan_without_pushdown_join_rule)
+
+    @patch("eva.catalog.catalog_manager.CatalogManager.get_udf_cost_catalog_entry")
+    def test_should_reorder_predicate(self, mock):
+        def side_effect_func(name):
+            if name == "DummyMultiObjectDetector":
+                return MagicMock(cost=10)
+            else:
+                return MagicMock(cost=5)
```
Comment (jarulraj) [inline test/integration_tests/test_optimizer_rules.py:135]: Should we have a test case for a scenario where there is no cost in the catalog? (defaults to `DEFAULT_FUNCTION_EXPRESSION_COST`)

**Review by jarulraj — APPROVED** [review]

File: test/benchmark_tests/test_benchmark_pytorch.py (hunk 3 above)
Comment (gaurav274) [reply test/benchmark_tests/test_benchmark_pytorch.py:83]: I changed the feature extractor to use a newer model.

File: test/integration_tests/test_optimizer_rules.py (hunk 4 above)
Comment (gaurav274) [reply test/integration_tests/test_optimizer_rules.py:135]: Done!

File: eva/constants.py (hunk 1 above)
Comment (gaurav274) [reply eva/constants.py:20]: Sure

File: eva/optimizer/rules/rules.py [hunk 5]
```diff
@@ -499,6 +503,64 @@ def apply(self, before: LogicalJoin, context: OptimizerContext):
         yield new_join
 
 
+class ReorderPredicates(Rule):
```
Comment (jiashenC) [inline eva/optimizer/rules/rules.py:506]: Should we give it a more specific name like `CostBasedReorder...` to differentiate with `SelectivityBasedReorder`?

File: eva/optimizer/rules/rules.py [hunk 6]
```diff
@@ -499,6 +503,64 @@ def apply(self, before: LogicalJoin, context: OptimizerContext):
...
+        return len(list(before.predicate.find_all(FunctionExpression))) > 0
+
+    def apply(self, before: LogicalFilter, context: OptimizerContext):
+        # Decompose the expression tree into a list of conjuncts
+        conjuncts = to_conjunction_list(before.predicate)
+
+        # Segregate the conjuncts into simple and function expressions
+        contains_func_exprs = []
+        simple_exprs = []
+        for conjunct in conjuncts:
```
Comment (jiashenC) [inline eva/optimizer/rules/rules.py:534]: Just make sure I understand this correctly, if  we have something like `OR(AND(a,b), c)`, this reorder rule won't be able to flip any order inside `OR` right?
